In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
addres

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

3 4
7

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Writing input1.txt


In [ ]:
%%writefile output1.txt
12

Writing output1.txt


In [ ]:
%%writefile input2.txt
0 9

Writing input2.txt


In [ ]:
%%writefile output2.txt
9

Writing output2.txt


In [ ]:
!./APlusB < input1.txt

/bin/bash: line 1: ./APlusB: No such file or directory


In [ ]:
!./APlusB < input2.txt

/bin/bash: line 1: ./APlusB: No such file or directory


### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt

In [ ]:
!grep -xvFf output1.txt output2.txt

9


In [ ]:
!grep -xvFf output2.txt output1.txt

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

/bin/bash: line 1: ./APlusB: No such file or directory


In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

/bin/bash: line 1: ./APlusB: No such file or directory


## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [ ]:
!python sum.py


^C


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [ ]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


In [ ]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

3
7 9 8
72


$$n^2 \ \text{first != second  comparisons},  n^2 - n \ \ \text{numbers[first] * numbers[second] products},  O(n^2)$$

### Test 1: Small numbers

In [4]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Writing input1.txt


In [ ]:
!./max_pairwise_product < input1.txt

/bin/bash: line 1: ./max_pairwise_product: No such file or directory


In [ ]:
%%writefile output1.txt
90

Writing output1.txt


In [ ]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Writing input2.txt


In [ ]:
%%writefile output2.txt
90

Writing output2.txt


In [ ]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

0


### Test 2: largest input numbers 200000 200000 in the input

In [ ]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [ ]:
200000*200000

40000000000

In [ ]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [ ]:
!./max_pairwise_product_full < input3.txt

/bin/bash: line 1: ./max_pairwise_product_full: No such file or directory


In [ ]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

/bin/bash: line 1: ./max_pairwise_product_full: No such file or directory


In [ ]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {1..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

cc1plus: fatal error: max_pairwise_product_full.cpp: No such file or directory
compilation terminated.
./run_all_test.sh: line 5: ./max_pairwise_product_full: No such file or directory
./run_all_test.sh: line 5: ./max_pairwise_product_full: No such file or directory
./run_all_test.sh: line 5: ./max_pairwise_product_full: No such file or directory


Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long_full.cpp


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [ ]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [ ]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

0


### Test 3  stress tests 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [ ]:
%%writefile input3.txt
6
1 3000 2000 9 20 23

Overwriting input3.txt


In [ ]:
%%writefile input4.txt
7
11 456 789 101112 131415 161718 192021

Writing input4.txt


In [ ]:
%%writefile input5.txt
5
951 753 654 852 379

Writing input5.txt


In [ ]:
%%writefile input6.txt
8
900 1000 1100 1050 850 660 780 222

Writing input6.txt


In [ ]:
%%writefile input7.txt
4
741 852 963 684

Writing input7.txt


In [ ]:
%%writefile input8.txt
5
167 943 381 927 200000

Writing input8.txt


In [ ]:
%%writefile input9.txt
6
1 3333 2000 9 2222 23

Writing input9.txt


In [ ]:
%%writefile input10.txt
6
1111 2222 3333 4444 5555 6666

Writing input10.txt


In [ ]:
!./max_pairwise_product_long_full < input1.txt
!./max_pairwise_product_long_full < input2.txt
!./max_pairwise_product_long_full < input3.txt > output3.txt
!./max_pairwise_product_long_full < input4.txt > output4.txt
!./max_pairwise_product_long_full < input5.txt > output5.txt
!./max_pairwise_product_long_full < input6.txt > output6.txt
!./max_pairwise_product_long_full < input7.txt > output7.txt
!./max_pairwise_product_long_full < input8.txt > output8.txt
!./max_pairwise_product_long_full < input9.txt > output9.txt
!./max_pairwise_product_long_full < input10.txt > output10.txt

90
0


### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and inoput12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [14]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
199999 * 200000

39999800000

In [12]:
%%writefile output11.txt
39999800000

Writing output11.txt


In [15]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
%%writefile output12.txt
39999800000

Writing output12.txt


Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

39999800000

real	4m56.497s
user	4m54.025s
sys	0m0.114s


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

39999800000

real	4m55.694s
user	4m53.500s
sys	0m0.080s


##### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //1+2n
        for (int second = 0; second < n; ++second) {                         //n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[1, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 4 + n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //N inst exec by line executed
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + n + [5*n**2+1, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  9 + 5*n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


Count executed intructions in code

In [40]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed

    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                //1+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                         //n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                               //n**2 -n
                    max_product = aux_product;                               //[1,n**2/2 - n/2]  upper bound
                    execints++;

                }
        }
      }
    }
                                                                            // --------------
    return max_product;                                                     // 4 + n + [5*n**2+1, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //N inst exec by line executed

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }



    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 4 + n + [5*n**2+1, 11*n**2/2  - n/2] upper bound


    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<"Number of instructions: " << execints << "\n";
    return 0;                                                // --------------
}                                                            // 9 + 5*n + [5*n**2+1, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt
!./max_pairwise_product_long_full_execints < input2.txt
!./max_pairwise_product_long_full_execints < input3.txt

589
9
221


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 9 + 5n + \left[5n^2+1, 11\frac{n^2}{2}  - \frac{n}{2}\right] \ \  \text{upper bound}$$

It depends on the size of the lists and  the order of the list.

For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 5(10) + 5(10)(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 589 $$

$$T(10) \in [560, 604]  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

   1  2  3  4  5  6  7  8  9  10
1  x  2  3  4  5  6  7  8  9  10         9
2  2  x  6  8  10 12 14 16 18 20         5 14
3  3  6  x  12 15 18 21 24 27 30         4 18
4  4  8  12 x  20 24 28 32 36 40         3 21
5  5  10 15 20 x  30 35 40 45 50         2 23
6  6  12 18 24 30 x  42 48 54 60         2 25
7  7  14 21 28 35 42 x  56 63 70         2 27
8  8  18 27 36 45 54 63 72 x  90         1 30
10 10 20 30 40 50 60 70 60 90 x          0

In [ ]:
9 + 5*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

589

In [ ]:
n = 10
print( 9 + 5*n + 5*n**2+1,  9 + 5*n + 11*n**2/2  - n/2)

560 604.0


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

589



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 9 + 5*(10) + 5*(10)**2 +  1  = 560 $$

In [ ]:
9 + 5*(10) + 5*(10)**2 +  1

560

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input4.txt
n= 10

304



For the input12.txt
```
200000
200000  199999 ... 2 1
```
$$T = 9 + 5*(200000) + 5*(200000)**2 +1  = 200001000010 $$



In [ ]:
9 + 5*(200000) + 5*(200000)**2 +1

200001000010

Calculate estimated time by instruction and number of instructions by second

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input12.txt


200001000021


In [ ]:
# Estimated time by instruction

306/200001000010


1.5299923499617505e-09

In [ ]:
# Estimated number of instructions by second

306/200001000010

1.5299923499617505e-09

# Estimate the maximun n that can be solved in 1 second

$$5n^2 \times 1.53*10^{-9} s = 1s$$

$$ n \approx 11000$$


In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.871s
user	0m0.862s
sys	0m0.002s


## Python

In [20]:
%%writefile max_pairwise_product_full.py
n = int(input())                        #1
list = input().split()                  #1
list = [int(x) for x in list]           #n

maxProduct = 0                          #1
auxProduct = 0                          #1
for i in range(0,n):                    #1+2n
  for j in range(0,n):                #n + 2n**2
    if i!=j:                            #n**2
      auxProduct = list[i]*list[j]      #n**2 -n
    if maxProduct < auxProduct:         #n**2 -n
      maxProduct = auxProduct           #[1, n**2/2 - n/2] upper bound

                                        # --------------
print(maxProduct)                       #5 + 3n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound

Writing max_pairwise_product_full.py


Testing with previous examples

In [ ]:
!chmod +x max_pairwise_product_full_py

!python max_pairwise_product_full.py < input3.txt > output3.txt
!python max_pairwise_product_full.py < input4.txt > output4.txt
!python max_pairwise_product_full.py < input5.txt > output5.txt
!python max_pairwise_product_full.py < input6.txt > output6.txt
!python max_pairwise_product_full.py < input7.txt > output7.txt
!python max_pairwise_product_full.py < input8.txt > output8.txt
!python max_pairwise_product_full.py < input9.txt > output9.txt
!python max_pairwise_product_full.py < input10.txt > output10.txt

Creating a big input

In [18]:
fileName = "inputPy.txt"

with open(fileName, "w") as file:
    file.write("20000"+ "\n")
    for i in range(1,20001):
        file.write(str(i) + " ")


Measuring the time

In [21]:
import time

start_time = time.time()
!python max_pairwise_product_full.py < inputPy.txt

end_time = time.time()
execution_time = end_time - start_time
print(f"Execution time: {execution_time} segs")
print(f"Execution time: {execution_time/60} min")

399980000
Execution time: 137.6328251361847 segs
Execution time: 2.2938804189364115 min


Program to count the number of steps it takes

In [37]:
%%writefile max_pairwise_product_full_count.py
op =0
n = int(input())                        #1
op+=1
list = input().split()                  #1
op+=1
list = [int(x) for x in list]           #n
op+=n


maxProduct = 0                          #1
op+=1
auxProduct = 0                          #1
op+=1
for i in range(0,n):                    #1+2n
  op+=1
  op+=2
  for j in range(0,n):                  #n + 2n**2
    op+=2
    op+=1
    if i!=j:                            #n**2
      auxProduct = list[i]*list[j]      #n**2 -n
      op+=1
      op+=1
    if maxProduct < auxProduct:         #n**2 -n
      maxProduct = auxProduct           #n * (n - 1) / 3
      op+=1

                                        # -------------- for worst case
print("Result: "+str(maxProduct))
print("Number of instructions: "+str(op))                       #5 + 3n + 3n**2 + 2(n**2 -n)+n * (n - 1) / 3

Overwriting max_pairwise_product_full_count.py


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T = 5 + 2n +5( n^{2}) +(n (n - 1) / 3)$$

It depends on the size of the lists and  the order of the list.

For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 5 + (20) + 5(10)(10) +  90/3= 555$$



In [28]:
!python max_pairwise_product_full_count.py < input1.txt

554


The number of instructions executed by the algooirithm with a list of size $n$
It depends on the size of the lists and  the order of the list.

For the inputPy.txt

$$T = 5 + (60000) + 3(20000)(20000) + 2(20000)(20000)-20000 + 199990/3= 2000106668$$



In [29]:
!python max_pairwise_product_full_count.py < inputPy.txt

2000261148


###Calculate estimated time by instruction and number of instructions by second

In [45]:
%%time
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

Number of instructions: 589
CPU times: user 19.7 ms, sys: 2.02 ms, total: 21.7 ms
Wall time: 916 ms


In [48]:
%%time
!python max_pairwise_product_full_count.py < input1.txt

Result: 90
Number of instructions: 554
CPU times: user 6.78 ms, sys: 0 ns, total: 6.78 ms
Wall time: 107 ms


In [47]:
time_per_ins = 0.0197 / 589
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction C++: {time_per_ins}")
print(f"Instructions per second C++: {instructions_per_second}")
time_per_ins = 0.00646 / 554
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction Python: {time_per_ins}")
print(f"Instructions per second Python: {instructions_per_second}")

Time per instruction C++: 3.344651952461799e-05
Instructions per second C++: 29898.477157360412
Time per instruction Python: 1.1660649819494584e-05
Instructions per second Python: 85758.51393188855


**If we replace T in the previous equations we have:**

Max n that c++ program can solve in a second: 74.5

Max n that python program can solve in a second: 126.64

## Perform the same analysis for an algorithm in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [10]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [19]:
!g++ max_pairwise_product.cpp -o max_pairwise_product
!time ./max_pairwise_product< input12.txt

2147483646

real	3m24.749s
user	3m20.039s
sys	0m0.083s


In [41]:
%%writefile max_pairwise_product_long_count.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed
    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1
    execints++;

    for (int first = 0; first < n; ++first) {                                //1+2n
        execints+=2;

        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n

        execints++;

			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
             execints++;
            execints++;
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;
                execints++;                                                  //[0, (n-1)n/2]
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] upper bound
}

int main() {                                              //N inst exec by line executed
    long long execints =0LL;
    int n;                                                //1
    execints++;

    std::cin >> n;                                        //1
    execints++;

    std::vector<int> numbers(n);                          //n
    execints+=n;

    for (int i = 0; i < n; ++i) {                         //1 + 2n
        execints+=2;

        std::cin >> numbers[i];                           //n
        execints++;
    }

    //std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout << "Number of instructions: " <<execints << "\n";
    return 0;                                             // --------------
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n] upper bound


Writing max_pairwise_product_long_count.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 9 + 7n + \left[2(n-1), 2.5(n-1)n\right] \ \  $$

It depends on the size of the lists and  the order of the list.

For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```

$$T(10) \in [97, 304]  $$


In [ ]:
n = 10
print(9 + 7*10+2*(n-1),9 + 7*10+2.5*(n-1)*n)

97 304.0


In [39]:
!g++ max_pairwise_product_long2.cpp -o max_pairwise_product_long2
!./max_pairwise_product_long2 < input1.txt

Number of instructions: 231


### Python

In [1]:
%%writefile max_pairwise_product.py
n = int(input())                        #1
list = input().split()                  #1
list = [int(x) for x in list]           #n

maxProduct = 0                          #1
auxProduct = 0                          #1
for i in range(0,n):                    #1+2n
  for j in range(i+1,n):                #((n**2)-n)/2
    if i!=j:                            #((n**2)-n)/2
      auxProduct = list[i]*list[j]      #((n**2)-n)/2
    if maxProduct < auxProduct:         #((n**2)-n)/2
      maxProduct = auxProduct           #[0,((n**2)-n)/2] upper bound

                                        # --------------
print(maxProduct)                       #5 + n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound

Writing max_pairwise_product.py


Testing our code with previous examples

In [ ]:
!chmod +x max_pairwise_product_py

!python max_pairwise_product.py < input3.txt > output3.txt
!python max_pairwise_product.py < input4.txt > output4.txt
!python max_pairwise_product.py < input5.txt > output5.txt
!python max_pairwise_product.py < input6.txt > output6.txt
!python max_pairwise_product.py < input7.txt > output7.txt
!python max_pairwise_product.py < input8.txt > output8.txt
!python max_pairwise_product.py < input9.txt > output9.txt
!python max_pairwise_product.py < input10.txt > output10.txt

In [3]:
import time

start_time = time.time()
!python max_pairwise_product.py < inputPy.txt

end_time = time.time()
execution_time = end_time - start_time
print(f"Execution time: {execution_time} segs")
print(f"Execution time: {execution_time/60} min")

399980000
Execution time: 63.087297677993774 segs
Execution time: 1.0514549612998962 min


Program to count the number of steps it takes

In [35]:
%%writefile max_pairwise_product_count.py
op = 0
xd =0
n = int(input())                        #1
op+=1
list = input().split()                  #1
op+=1
list = [int(x) for x in list]           #n
op += n

maxProduct = 0                          #1
op+=1
auxProduct = 0                          #1
op+=1
for i in range(0,n):                    #1+2n
  op+=1
  op+=2
  for j in range(i+1,n):                #((n**2)-n)/2
    op+=1
    if i!=j:                            #((n**2)-n)/2
      auxProduct = list[i]*list[j]      #((n**2)-n)/2
      op+=1
      op+=1
    op+=1
    if maxProduct < auxProduct:         #n * (n - 1) / 2
      maxProduct = auxProduct           #n * (n - 1) / 3
      op+=1

                                        # --------------For worst case
print("Result: "+str(maxProduct))
print("Number of instructions: "+str(op))                       #5 + 4n +2( n * (n - 1)) + (n * (n - 1) / 3)

Overwriting max_pairwise_product_count.py


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T = 5 + 4n +2( n(n - 1)) + (n (n - 1) / 3)$$

It depends on the size of the lists and  the order of the list.

For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 5 + (40) + 2(10)(9) + (90)/3  = 259$$



In [6]:
!python max_pairwise_product_count.py < input1.txt

254


For a big input

The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T = 5 + 4n +2( n(n - 1)) + (n (n - 1) / 3)$$

It depends on the size of the lists and  the order of the list.

For the inputPy.txt with 20000 numbers

$$T = 5 + (80000) + 2(20000)(19999) + (199990)/3  = 800106668$$



In [7]:
!python max_pairwise_product_count.py < inputPy.txt

800261148


###Calculate estimated time by instruction and number of instructions by second

In [43]:
%%time
!g++ max_pairwise_product_long_count.cpp -o max_pairwise_product_long_count
!./max_pairwise_product_long_count < input1.txt

Number of instructions: 231
CPU times: user 13.4 ms, sys: 2.93 ms, total: 16.3 ms
Wall time: 813 ms


In [55]:
%%time
!python max_pairwise_product_count.py < input1.txt

Result: 90
Number of instructions: 254
CPU times: user 4.9 ms, sys: 1.01 ms, total: 5.91 ms
Wall time: 105 ms


In [51]:
time_per_ins = 0.0134 / 231
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction C++: {time_per_ins}")
print(f"Instructions per second C++: {instructions_per_second}")
time_per_ins = 0.0045 / 254
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction Python: {time_per_ins}")
print(f"Instructions per second Python: {instructions_per_second}")

Time per instruction C++: 5.800865800865801e-05
Instructions per second C++: 17238.805970149253
Time per instruction Python: 1.7716535433070864e-05
Instructions per second Python: 56444.44444444445


**If we replace T in the previous equations we have:**

Max n that c++ program can solve in a second: 82.12

Max n that python program can solve in a second: 154.95

### Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile max_pairwise_sort_nlogn.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    cout << "C++" << "\n";
    int n;                                     // 1
    cin >> n;                                  // 1
    vector<int> numbers(n);                    // n

    for (int i = 0; i < n; ++i) {              //1 + 2n
        cin >> numbers[i];                     // 1
    }

    sort(numbers.begin(), numbers.end());      // n*long
    cout <<  "Product: " << numbers[numbers.size() - 1] * numbers[numbers.size() - 2] << "\n"; // 1
    return 0;                                  // 1
}

Writing max_pairwise_sort_nlogn.cpp


In [ ]:
%%writefile max_pairwise_sort_nlogn.py
print("Python 3")
# Read input values
n = int(input())              # 1
numbers = list(map(int, input().split()))  # n

# Sort the list
numbers.sort()                # n*log(n)

# Calculate the maximum pairwise product
result = numbers[-1] * numbers[-2]  # 1

# Print the result
print(f"Product: {result}")

Writing max_pairwise_sort_nlogn.py


Test with inputs 1 and 2



In [ ]:
!g++ max_pairwise_sort_nlogn.cpp -o max_pairwise_sort_nlogn
!./max_pairwise_sort_nlogn < input1.txt
!./max_pairwise_sort_nlogn < input2.txt
!python max_pairwise_sort_nlogn.py < input1.txt
!python max_pairwise_sort_nlogn.py < input2.txt

C++
Product: 90
C++
Product: 90
Python 3
Product: 90
Python 3
Product: 90


Test and time it with input 3

In [ ]:
!g++ max_pairwise_sort_nlogn.cpp -o max_pairwise_sort_nlogn
!./max_pairwise_sort_nlogn < input3.txt

C++
Product: 6000000


20 stress tests

In [ ]:
#CREATE 20 random texts
import random

for i in range(1,21):
    fileName = f"stress_input_{i}.txt"

    with open(fileName, "w") as file:
      n = random.randint(3, 40)
      file.write(f"{n}"+ "\n")
      for i in range(1,n):
          file.write(str(random.randint(1, 100)) + " ")


In [ ]:
!g++ max_pairwise_sort_nlogn.cpp -o max_pairwise_sort_nlogn
!./max_pairwise_sort_nlogn < stress_input_1.txt
!./max_pairwise_sort_nlogn < stress_input_2.txt
!./max_pairwise_sort_nlogn < stress_input_3.txt
!./max_pairwise_sort_nlogn < stress_input_4.txt
!./max_pairwise_sort_nlogn < stress_input_5.txt
!./max_pairwise_sort_nlogn < stress_input_6.txt
!./max_pairwise_sort_nlogn < stress_input_7.txt
!./max_pairwise_sort_nlogn < stress_input_8.txt
!./max_pairwise_sort_nlogn < stress_input_9.txt
!./max_pairwise_sort_nlogn < stress_input_10.txt
!./max_pairwise_sort_nlogn < stress_input_11.txt
!./max_pairwise_sort_nlogn < stress_input_12.txt
!./max_pairwise_sort_nlogn < stress_input_13.txt
!./max_pairwise_sort_nlogn < stress_input_14.txt
!./max_pairwise_sort_nlogn < stress_input_15.txt
!./max_pairwise_sort_nlogn < stress_input_16.txt
!./max_pairwise_sort_nlogn < stress_input_17.txt
!./max_pairwise_sort_nlogn < stress_input_18.txt
!./max_pairwise_sort_nlogn < stress_input_19.txt
!./max_pairwise_sort_nlogn < stress_input_20.txt


C++
Product: 4380
C++
Product: 9120
C++
Product: 8930
C++
Product: 8742
C++
Product: 7636
C++
Product: 9600
C++
Product: 9702
C++
Product: 6241
C++
Product: 8928
C++
Product: 5850
C++
Product: 9405
C++
Product: 9500
C++
Product: 5100
C++
Product: 9306
C++
Product: 7476
C++
Product: 9120
C++
Product: 7308
C++
Product: 9016
C++
Product: 8613
C++
Product: 9500


In [ ]:
!python max_pairwise_sort_nlogn.py < stress_input_1.txt
!python max_pairwise_sort_nlogn.py < stress_input_2.txt
!python max_pairwise_sort_nlogn.py < stress_input_3.txt
!python max_pairwise_sort_nlogn.py < stress_input_4.txt
!python max_pairwise_sort_nlogn.py < stress_input_5.txt
!python max_pairwise_sort_nlogn.py < stress_input_6.txt
!python max_pairwise_sort_nlogn.py < stress_input_7.txt
!python max_pairwise_sort_nlogn.py < stress_input_8.txt
!python max_pairwise_sort_nlogn.py < stress_input_9.txt
!python max_pairwise_sort_nlogn.py < stress_input_10.txt
!python max_pairwise_sort_nlogn.py < stress_input_11.txt
!python max_pairwise_sort_nlogn.py < stress_input_12.txt
!python max_pairwise_sort_nlogn.py < stress_input_13.txt
!python max_pairwise_sort_nlogn.py < stress_input_14.txt
!python max_pairwise_sort_nlogn.py < stress_input_15.txt
!python max_pairwise_sort_nlogn.py < stress_input_16.txt
!python max_pairwise_sort_nlogn.py < stress_input_17.txt
!python max_pairwise_sort_nlogn.py < stress_input_18.txt
!python max_pairwise_sort_nlogn.py < stress_input_19.txt
!python max_pairwise_sort_nlogn.py < stress_input_20.txt

Python 3
Product: 4380
Python 3
Product: 9120
Python 3
Product: 8930
Python 3
Product: 8742
Python 3
Product: 7636
Python 3
Product: 9600
Python 3
Product: 9702
Python 3
Product: 6241
Python 3
Product: 8928
Python 3
Product: 5850
Python 3
Product: 9405
Python 3
Product: 9500
Python 3
Product: 5100
Python 3
Product: 9306
Python 3
Product: 7476
Python 3
Product: 9120
Python 3
Product: 7308
Python 3
Product: 9016
Python 3
Product: 8613
Python 3
Product: 9500


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile max_pairwise_sort_nlogn_count.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <bits/stdc++.h>

using namespace std;

int main() {
    cout << "C++" << "\n";

    uint64_t execints = 0;

    int n;                                     // 1
    execints++;
    cin >> n;                                  // 1
    execints++;
    vector<int> numbers(n);                    // n
    execints += n;

    for (int i = 0; i < n; ++i) {              //1 + 2n
        execints += 2;
        cin >> numbers[i];                     // 1
        execints++;
    }

    sort(numbers.begin(), numbers.end());      // n*long

    execints += n * log2(n);

    cout <<  numbers[n - 1] * numbers[n - 2] << "\n"; // 3

    execints += 3;

    cout << "Executed instructions: " << execints << "\n";

    return 0;                                  // 1
}

Writing max_pairwise_sort_nlogn_count.cpp


In [ ]:
%%writefile max_pairwise_sort_nlogn_count.py

import math

print("Python 3")

# Read input values

execints = 0
n = int(input())              # 1

execints += 1
numbers = list(map(int, input().split()))  # 4n (n input read + map o(n) + create list + cast to int values)

execints += 4 * n

# Sort the list
numbers.sort()                # n*log(n)

execints += math.log2(n)

# Calculate the maximum pairwise product
result = numbers[-1] * numbers[-2]  # 3

execints += 3

# Print the result
print(f"Executed instructions: {int(execints)}")

Overwriting max_pairwise_sort_nlogn_count.py


In [ ]:
!g++ max_pairwise_sort_nlogn_count.cpp -o max_pairwise_sort_nlogn_count
!./max_pairwise_sort_nlogn_count < input1.txt
!./max_pairwise_sort_nlogn_count < input2.txt
!./max_pairwise_sort_nlogn_count < input3.txt
!python max_pairwise_sort_nlogn_count.py < input1.txt
!python max_pairwise_sort_nlogn_count.py < input2.txt
!python max_pairwise_sort_nlogn_count.py < input3.txt

C++
90
Executed instructions: 78
C++
90
Executed instructions: 78
C++
6000000
Executed instructions: 44
Python 3
Executed instructions: 47
Python 3
Executed instructions: 47
Python 3
Executed instructions: 30


Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%time
!g++ max_pairwise_sort_nlogn_count.cpp -o max_pairwise_sort_nlogn_count
!./max_pairwise_sort_nlogn_count < input1.txt

C++
90
Executed instructions: 78
CPU times: user 24.3 ms, sys: 2.12 ms, total: 26.4 ms
Wall time: 1.92 s


In [ ]:
%%time
!python max_pairwise_sort_nlogn_count.py < input1.txt

Python 3
Executed instructions: 47
CPU times: user 5.14 ms, sys: 1.97 ms, total: 7.11 ms
Wall time: 109 ms


In [ ]:
time_per_ins = 0.0243 / 78
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction C++: {time_per_ins}")
print(f"Instructions per second C++: {instructions_per_second}")
time_per_ins = 0.00514 / 47
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction Python: {time_per_ins}")
print(f"Instructions per second Python: {instructions_per_second}")

Time per instruction C++: 0.0003115384615384615
Instructions per second C++: 3209.8765432098767
Time per instruction Python: 0.00010936170212765956
Instructions per second Python: 9143.968871595333


### Efficient solution largest and second-largest in an list, array or vector in C++ and Python
 $$2.5 n = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile max_pairwise_sort_efficient.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits.h>

using namespace std;

int main() {
    cout << "--- C++ ---" << "\n";
    int n;                                     // 1
    cin >> n;                                  // 1
    int maxNumber = INT_MIN;                   // 1
    int secondMaxNumber = INT_MIN;             // 1

    vector<int> numbers(n);                    // n

    for (int i = 0; i < n; ++i) {              // 1 + 2n
        cin >> numbers[i];                     // n
    }


    for (int i = 0; i < n; ++i) {              //1 + 2n
        int aux = numbers[i];                           // n

        if (aux > maxNumber){                  // n
          secondMaxNumber = maxNumber;         // [0, n] in the worst case it would be n times
          maxNumber = aux;                     // [0, n] in the worst case it would be n times

        } else if (aux > secondMaxNumber) {    // n
          secondMaxNumber = aux;               // [0, n] in the worst case it would be n times
        }

    }

    cout <<  "Product: " << maxNumber * secondMaxNumber << "\n"; // 1
    return 0;                                  // 1
}

Overwriting max_pairwise_sort_efficient.cpp


In [ ]:
%%writefile max_pairwise_sort_efficient.py
print("--- Python ---")
n = int(input())
numbers = list(map(int, input().split()))

max_number = float("-inf")                   # 1
second_max_number = float("-inf")            # 1

for num in numbers:                          # 1 + 2n
    if num > max_number:                     # n
        second_max_number = max_number       # [0, n] in the worst case it would be n times
        max_number = num                     # [0, n] in the worst case it would be n times
    elif num > second_max_number:            # n
        second_max_number = num              # [0, n] in the worst case it would be n times

product = max_number * second_max_number     # 1
print(f"Product: {product}")                 # 1

Overwriting max_pairwise_sort_efficient.py


Test with inputs 1 and 2



In [ ]:
!g++ max_pairwise_sort_efficient.cpp -o max_pairwise_sort_efficient
!./max_pairwise_sort_efficient < input1.txt
!./max_pairwise_sort_efficient < input2.txt
! python max_pairwise_sort_efficient.py < input1.txt
! python max_pairwise_sort_efficient.py < input2.txt

--- C++ ---
Product: 90
--- C++ ---
Product: 90
--- Python ---
Product: 90
--- Python ---
Product: 90


Test and time it with input 3

In [ ]:
!g++ max_pairwise_sort_efficient.cpp -o max_pairwise_sort_efficient
!./max_pairwise_sort_efficient < input3.txt
! python max_pairwise_sort_efficient.py < input3.txt

--- C++ ---
Product: 6000000
--- Python ---
Product: 6000000


***20*** stress tests

In [ ]:
#CREATE 20 random texts
import random

for i in range(1,21):
    fileName = f"stress_input_{i}.txt"

    with open(fileName, "w") as file:
      n = random.randint(3, 40)
      file.write(f"{n}"+ "\n")
      for i in range(1,n):
          file.write(str(random.randint(1, 100)) + " ")

In [ ]:
!g++ max_pairwise_sort_efficient.cpp -o max_pairwise_sort_efficient
!./max_pairwise_sort_efficient < stress_input_1.txt
!./max_pairwise_sort_efficient < stress_input_2.txt
!./max_pairwise_sort_efficient < stress_input_3.txt
!./max_pairwise_sort_efficient < stress_input_4.txt
!./max_pairwise_sort_efficient < stress_input_5.txt
!./max_pairwise_sort_efficient < stress_input_6.txt
!./max_pairwise_sort_efficient < stress_input_7.txt
!./max_pairwise_sort_efficient < stress_input_8.txt
!./max_pairwise_sort_efficient < stress_input_9.txt
!./max_pairwise_sort_efficient < stress_input_10.txt
!./max_pairwise_sort_efficient < stress_input_11.txt
!./max_pairwise_sort_efficient < stress_input_12.txt
!./max_pairwise_sort_efficient < stress_input_13.txt
!./max_pairwise_sort_efficient < stress_input_14.txt
!./max_pairwise_sort_efficient < stress_input_15.txt
!./max_pairwise_sort_efficient < stress_input_16.txt
!./max_pairwise_sort_efficient < stress_input_17.txt
!./max_pairwise_sort_efficient < stress_input_18.txt
!./max_pairwise_sort_efficient < stress_input_19.txt
!./max_pairwise_sort_efficient < stress_input_20.txt

--- C++ ---
Product: 8924
--- C++ ---
Product: 6048
--- C++ ---
Product: 9207
--- C++ ---
Product: 9604
--- C++ ---
Product: 9216
--- C++ ---
Product: 5369
--- C++ ---
Product: 9800
--- C++ ---
Product: 8742
--- C++ ---
Product: 9700
--- C++ ---
Product: 6006
--- C++ ---
Product: 7644
--- C++ ---
Product: 8928
--- C++ ---
Product: 8645
--- C++ ---
Product: 6320
--- C++ ---
Product: 4331
--- C++ ---
Product: 9306
--- C++ ---
Product: 8439
--- C++ ---
Product: 8170
--- C++ ---
Product: 9405
--- C++ ---
Product: 7426


In [ ]:
! python max_pairwise_sort_efficient.py < stress_input_1.txt
! python max_pairwise_sort_efficient.py < stress_input_2.txt
! python max_pairwise_sort_efficient.py < stress_input_3.txt
! python max_pairwise_sort_efficient.py < stress_input_4.txt
! python max_pairwise_sort_efficient.py < stress_input_5.txt
! python max_pairwise_sort_efficient.py < stress_input_6.txt
! python max_pairwise_sort_efficient.py < stress_input_7.txt
! python max_pairwise_sort_efficient.py < stress_input_8.txt
! python max_pairwise_sort_efficient.py < stress_input_9.txt
! python max_pairwise_sort_efficient.py < stress_input_10.txt
! python max_pairwise_sort_efficient.py < stress_input_11.txt
! python max_pairwise_sort_efficient.py < stress_input_12.txt
! python max_pairwise_sort_efficient.py < stress_input_13.txt
! python max_pairwise_sort_efficient.py < stress_input_14.txt
! python max_pairwise_sort_efficient.py < stress_input_15.txt
! python max_pairwise_sort_efficient.py < stress_input_16.txt
! python max_pairwise_sort_efficient.py < stress_input_17.txt
! python max_pairwise_sort_efficient.py < stress_input_18.txt
! python max_pairwise_sort_efficient.py < stress_input_19.txt
! python max_pairwise_sort_efficient.py < stress_input_20.txt

--- Python ---
Product: 8924
--- Python ---
Product: 6048
--- Python ---
Product: 9207
--- Python ---
Product: 9604
--- Python ---
Product: 9216
--- Python ---
Product: 5369
--- Python ---
Product: 9800
--- Python ---
Product: 8742
--- Python ---
Product: 9700
--- Python ---
Product: 6006
--- Python ---
Product: 7644
--- Python ---
Product: 8928
--- Python ---
Product: 8645
--- Python ---
Product: 6320
--- Python ---
Product: 4331
--- Python ---
Product: 9306
--- Python ---
Product: 8439
--- Python ---
Product: 8170
--- Python ---
Product: 9405
--- Python ---
Product: 7426


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile max_pairwise_sort_efficient_count.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits.h>

using namespace std;

int main() {
    cout << "--- C++ ---" << "\n";
    int exexins = 0;
    int n;                                     // 1
    exexins++;
    cin >> n;                                  // 1
    exexins++;
    int maxNumber = INT_MIN;                   // 1
    exexins++;
    int secondMaxNumber = INT_MIN;             // 1
    exexins++;

    vector<int> numbers(n);                    // n

    exexins += n;

    for (int i = 0; i < n; ++i) {              // 1 + 2n
        cin >> numbers[i];                     // n
        exexins++;
    }


    exexins++;
    for (int i = 0; i < n; ++i) {              //1 + 2n
        exexins += 2;

        int aux = numbers[i];                           // n
        exexins++;


        exexins++; // if
        exexins++; // else if
        if (aux > maxNumber){                  // n
          secondMaxNumber = maxNumber;         // [0, n] in the worst case it would be n times
          exexins++;
          maxNumber = aux;                     // [0, n] in the worst case it would be n times
          exexins++;
        } else if (aux > secondMaxNumber) {    // n
          secondMaxNumber = aux;               // [0, n] in the worst case it would be n times
          exexins++;
        }

    }

    cout <<  "Executed instructions: " << exexins << " n: "<< n << " Product: " << maxNumber * secondMaxNumber << "\n"; // 1
    return 0;                                  // 1
}

Overwriting max_pairwise_sort_efficient_count.cpp


In [ ]:
!g++ max_pairwise_sort_efficient_count.cpp -o max_pairwise_sort_efficient_count
!./max_pairwise_sort_efficient_count < input1.txt
!./max_pairwise_sort_efficient_count < input2.txt
!./max_pairwise_sort_efficient_count < input3.txt

--- C++ ---
Executed instructions: 95 n: 10 Product: 90
--- C++ ---
Executed instructions: 78 n: 10 Product: 90
--- C++ ---
Executed instructions: 52 n: 6 Product: 6000000


In [ ]:
%%writefile max_pairwise_sort_efficient_count.py
print("--- Python ---")
exexins = 0
n = int(input())                             # 1
exexins += 1
numbers = list(map(int, input().split()))    # 4n (n input read + map o(n) + create list + cast to int values)
exexins = 4 * n

max_number = float("-inf")                   # 1
exexins += 1
second_max_number = float("-inf")            # 1
exexins += 1

exexins += 1
for num in numbers:                          # 1 + 2n
    exexins += 2

    exexins += 1 # if
    exexins += 1 # elif
    if num > max_number:                     # n
        second_max_number = max_number       # [0, n] in the worst case it would be n times
        exexins += 1
        max_number = num                     # [0, n] in the worst case it would be n times
        exexins += 1
    elif num > second_max_number:            # n
        second_max_number = num              # [0, n] in the worst case it would be n times
        exexins += 1

product = max_number * second_max_number     # 1
exexins += 1
print(f"Executed instructions: {exexins} n: {n} Product: {product}")                 # 1


Overwriting max_pairwise_sort_efficient_count.py


In [ ]:
!python max_pairwise_sort_efficient_count.py < input1.txt
!python max_pairwise_sort_efficient_count.py < input2.txt
!python max_pairwise_sort_efficient_count.py < input3.txt

--- Python ---
Executed instructions: 104 n: 10 Product: 90
--- Python ---
Executed instructions: 87 n: 10 Product: 90
--- Python ---
Executed instructions: 57 n: 6 Product: 6000000


Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%time
!g++ max_pairwise_sort_efficient_count.cpp -o max_pairwise_sort_efficient_count
!./max_pairwise_sort_efficient_count < input1.txt

--- C++ ---
Executed instructions: 95 n: 10 Product: 90
CPU times: user 11.5 ms, sys: 781 µs, total: 12.3 ms
Wall time: 509 ms


In [ ]:
%%time
!python max_pairwise_sort_efficient_count.py < input1.txt

--- Python ---
Executed instructions: 104 n: 10 Product: 90
CPU times: user 2.9 ms, sys: 1.14 ms, total: 4.05 ms
Wall time: 104 ms


In [ ]:
time_per_ins = 0.0115 / 85
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction C++: {time_per_ins}")
print(f"Instructions per second C++: {instructions_per_second}")
time_per_ins = 0.0029 / 104
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction Python: {time_per_ins}")
print(f"Instructions per second Python: {instructions_per_second}")

Time per instruction C++: 0.0001352941176470588
Instructions per second C++: 7391.304347826088
Time per instruction Python: 2.7884615384615383e-05
Instructions per second Python: 35862.06896551725


### Super efficient solution IN C++ AND PYTHON (both)

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python
 $$2 n = O(n)$$

In [ ]:
%%writefile max_pairwise_sort_super_efficient.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits.h>

using namespace std;

int main() {
    cout << "--- C++ ---" << "\n";
    int n;                                     // 1
    cin >> n;                                  // 1
    int maxNumber = INT_MIN;                   // 1
    int secondMaxNumber = INT_MIN;             // 1


    for (int i = 0; i < n; ++i) {              //1 + 2n
        int aux = 0;                           // n
        cin >> aux;                            // n

        if (aux > maxNumber){                  // n
          secondMaxNumber = maxNumber;         // [0, n] in the worst case it would be n times
          maxNumber = aux;                     // [0, n] in the worst case it would be n times

        } else if (aux > secondMaxNumber) {    // n
          secondMaxNumber = aux;               // [0, n] in the worst case it would be n times
        }

    }

    cout <<  "Product: " << maxNumber * secondMaxNumber << "\n"; // 1
    return 0;                                  // 1
}

Overwriting max_pairwise_sort_super_efficient.cpp


In [ ]:
%%writefile max_pairwise_sort_super_efficient.py
print("--- Python ---")
n = int(input())                            # 1
max_number = float("-inf")                  # 1
second_max_number = float("-inf")           # 1

for num in input().split():                 # 3 + 2n

    num = int(num)                          # n

    if num > max_number:                    # n
        second_max_number = max_number      # [0, n] in the worst case it would be n times
        max_number = num                    # [0, n] in the worst case it would be n times
    elif num > second_max_number:           # n
        second_max_number = num             # [0, n] in the worst case it would be n times

print(f"Product: {max_number * second_max_number}")

Overwriting max_pairwise_sort_super_efficient.py


Test with inputs 1 and 2




In [ ]:
!g++ max_pairwise_sort_super_efficient.cpp -o max_pairwise_sort_super_efficient
!./max_pairwise_sort_super_efficient < input1.txt
!./max_pairwise_sort_super_efficient < input2.txt
!python max_pairwise_sort_super_efficient.py < input1.txt
!python max_pairwise_sort_super_efficient.py < input2.txt

--- C++ ---
Product: 90
--- C++ ---
Product: 90
--- Python ---
Product: 90
--- Python ---
Product: 90


Test and time it with input 3

In [ ]:
!g++ max_pairwise_sort_super_efficient.cpp -o max_pairwise_sort_super_efficient
!./max_pairwise_sort_super_efficient < input3.txt
!python max_pairwise_sort_super_efficient.py < input3.txt

--- C++ ---
Product: 6000000
--- Python ---
Product: 6000000


20 stress tests

In [ ]:
#CREATE 20 random texts
import random

for i in range(1,21):
    fileName = f"stress_input_{i}.txt"

    with open(fileName, "w") as file:
      n = random.randint(3, 40)
      file.write(f"{n}"+ "\n")
      for i in range(1,n):
          file.write(str(random.randint(1, 100)) + " ")

In [ ]:
!g++ max_pairwise_sort_super_efficient.cpp -o max_pairwise_sort_super_efficient
!./max_pairwise_sort_super_efficient < stress_input_1.txt
!./max_pairwise_sort_super_efficient < stress_input_2.txt
!./max_pairwise_sort_super_efficient < stress_input_3.txt
!./max_pairwise_sort_super_efficient < stress_input_4.txt
!./max_pairwise_sort_super_efficient < stress_input_5.txt
!./max_pairwise_sort_super_efficient < stress_input_6.txt
!./max_pairwise_sort_super_efficient < stress_input_7.txt
!./max_pairwise_sort_super_efficient < stress_input_8.txt
!./max_pairwise_sort_super_efficient < stress_input_9.txt
!./max_pairwise_sort_super_efficient < stress_input_10.txt
!./max_pairwise_sort_super_efficient < stress_input_11.txt
!./max_pairwise_sort_super_efficient < stress_input_12.txt
!./max_pairwise_sort_super_efficient < stress_input_13.txt
!./max_pairwise_sort_super_efficient < stress_input_14.txt
!./max_pairwise_sort_super_efficient < stress_input_15.txt
!./max_pairwise_sort_super_efficient < stress_input_16.txt
!./max_pairwise_sort_super_efficient < stress_input_17.txt
!./max_pairwise_sort_super_efficient < stress_input_18.txt
!./max_pairwise_sort_super_efficient < stress_input_19.txt
!./max_pairwise_sort_super_efficient < stress_input_20.txt

--- C++ ---
Product: 5950
--- C++ ---
Product: 8265
--- C++ ---
Product: 7790
--- C++ ---
Product: 9408
--- C++ ---
Product: 8178
--- C++ ---
Product: 8811
--- C++ ---
Product: 9702
--- C++ ---
Product: 8360
--- C++ ---
Product: 10000
--- C++ ---
Product: 9800
--- C++ ---
Product: 5041
--- C++ ---
Product: 8272
--- C++ ---
Product: 9603
--- C++ ---
Product: 8188
--- C++ ---
Product: 9900
--- C++ ---
Product: 9604
--- C++ ---
Product: 8455
--- C++ ---
Product: 9801
--- C++ ---
Product: 8372
--- C++ ---
Product: 8366


In [ ]:
!python max_pairwise_sort_super_efficient.py < stress_input_1.txt
!python max_pairwise_sort_super_efficient.py < stress_input_2.txt
!python max_pairwise_sort_super_efficient.py < stress_input_3.txt
!python max_pairwise_sort_super_efficient.py < stress_input_4.txt
!python max_pairwise_sort_super_efficient.py < stress_input_5.txt
!python max_pairwise_sort_super_efficient.py < stress_input_6.txt
!python max_pairwise_sort_super_efficient.py < stress_input_7.txt
!python max_pairwise_sort_super_efficient.py < stress_input_8.txt
!python max_pairwise_sort_super_efficient.py < stress_input_9.txt
!python max_pairwise_sort_super_efficient.py < stress_input_10.txt
!python max_pairwise_sort_super_efficient.py < stress_input_11.txt
!python max_pairwise_sort_super_efficient.py < stress_input_12.txt
!python max_pairwise_sort_super_efficient.py < stress_input_13.txt
!python max_pairwise_sort_super_efficient.py < stress_input_14.txt
!python max_pairwise_sort_super_efficient.py < stress_input_15.txt
!python max_pairwise_sort_super_efficient.py < stress_input_16.txt
!python max_pairwise_sort_super_efficient.py < stress_input_17.txt
!python max_pairwise_sort_super_efficient.py < stress_input_18.txt
!python max_pairwise_sort_super_efficient.py < stress_input_19.txt
!python max_pairwise_sort_super_efficient.py < stress_input_20.txt

--- Python ---
Product: 5950
--- Python ---
Product: 8265
--- Python ---
Product: 7790
--- Python ---
Product: 9408
--- Python ---
Product: 8178
--- Python ---
Product: 8811
--- Python ---
Product: 9702
--- Python ---
Product: 8360
--- Python ---
Product: 7500
--- Python ---
Product: 9800
--- Python ---
Product: 3905
--- Python ---
Product: 8272
--- Python ---
Product: 9603
--- Python ---
Product: 8188
--- Python ---
Product: 9900
--- Python ---
Product: 9604
--- Python ---
Product: 8455
--- Python ---
Product: 9801
--- Python ---
Product: 8372
--- Python ---
Product: 8366


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile max_pairwise_sort_super_efficient_count.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits.h>

using namespace std;

int main() {
    cout << "--- C++ ---" << "\n";
    int exexins = 0;
    int n;                                     // 1
    exexins++;
    cin >> n;                                  // 1
    exexins++;
    int maxNumber = INT_MIN;                   // 1
    exexins++;
    int secondMaxNumber = INT_MIN;             // 1
    exexins++;


    exexins++;
    for (int i = 0; i < n; ++i) {              //1 + 2n
        exexins += 2;
        int aux = 0;                           // n
        exexins++;

        cin>> aux;                            // n
        exexins++;


        exexins++; // if
        exexins++; // else if
        if (aux > maxNumber){                  // n
          secondMaxNumber = maxNumber;         // [0, n] in the worst case it would be n times
          exexins++;
          maxNumber = aux;                     // [0, n] in the worst case it would be n times
          exexins++;
        } else if (aux > secondMaxNumber) {    // n
          secondMaxNumber = aux;               // [0, n] in the worst case it would be n times
          exexins++;
        }

    }

    cout <<  "Executed instructions: " << exexins << " n: "<< n << " Product: " << maxNumber * secondMaxNumber << "\n"; // 1
    return 0;                                  // 1
}

Writing max_pairwise_sort_super_efficient_count.cpp


In [ ]:
%%writefile max_pairwise_sort_super_efficient_count.py
print("--- Python ---")
exexins = 0
n = int(input())                            # 1
exexins += 1
max_number = float("-inf")                  # 1
exexins += 1
second_max_number = float("-inf")           # 1
exexins += 1

exexins += 3
for num in input().split():                 # 3 + 2n
    exexins += 2
    num = int(num)                          # n
    exexins += 1

    exexins += 1 # if
    exexins += 1 # elif
    if num > max_number:                    # n
        second_max_number = max_number      # [0, n] in the worst case it would be n times
        exexins += 1
        max_number = num                    # [0, n] in the worst case it would be n times
        exexins += 1
    elif num > second_max_number:           # n
        second_max_number = num             # [0, n] in the worst case it would be n times
        exexins += 1

print(f"Executed instructions: {exexins} n: {n} Product: {max_number * second_max_number}")

Overwriting max_pairwise_sort_super_efficient_count.py


In [ ]:
!g++ max_pairwise_sort_super_efficient_count.cpp -o max_pairwise_sort_super_efficient_count
!./max_pairwise_sort_super_efficient_count < input1.txt
!./max_pairwise_sort_super_efficient_count < input2.txt
!./max_pairwise_sort_super_efficient_count < input3.txt
!python max_pairwise_sort_super_efficient_count.py < input1.txt
!python max_pairwise_sort_super_efficient_count.py < input2.txt
!python max_pairwise_sort_super_efficient_count.py < input3.txt

--- C++ ---
Executed instructions: 85 n: 10 Product: 90
--- C++ ---
Executed instructions: 68 n: 10 Product: 90
--- C++ ---
Executed instructions: 46 n: 6 Product: 6000000
--- Python ---
Executed instructions: 76 n: 10 Product: 90
--- Python ---
Executed instructions: 59 n: 10 Product: 90
--- Python ---
Executed instructions: 41 n: 6 Product: 6000000


Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%time
!g++ max_pairwise_sort_super_efficient_count.cpp -o max_pairwise_sort_super_efficient_count
!./max_pairwise_sort_super_efficient_count < input1.txt

--- C++ ---
Executed instructions: 85 n: 10 Product: 90
CPU times: user 11 ms, sys: 811 µs, total: 11.8 ms
Wall time: 709 ms


In [ ]:
%%time
!python max_pairwise_sort_super_efficient_count.py < input1.txt

--- Python ---
Executed instructions: 76 n: 10 Product: 90
CPU times: user 3.61 ms, sys: 1.09 ms, total: 4.7 ms
Wall time: 104 ms


In [ ]:
time_per_ins = 0.011 / 85
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction C++: {time_per_ins}")
print(f"Instructions per second C++: {instructions_per_second}")
time_per_ins = 0.00361 / 76
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction Python: {time_per_ins}")
print(f"Instructions per second Python: {instructions_per_second}")

Time per instruction C++: 0.00012941176470588234
Instructions per second C++: 7727.272727272728
Time per instruction Python: 4.7499999999999996e-05
Instructions per second Python: 21052.63157894737
